### PR TITLE
Generate docs at build time instead of install time

### PR DESCRIPTION
**Pull request by levitte**

### CODE REVIEW

**Comment by dot-asm** [comment]
Here is a thing (aside CI problems that is). If you rebuild often from scratch (I do it a lot with different configs and compilers), it's only appropriate if build time is as minimal as possible. And it becomes real annoyance and even problem on low-end systems with poor i/o, such as Raspberry Pi. Yes, one can say "call make build_sw", but one can just as well say "why don't call make build_docs". Basically it's about active developer's experience. And question becomes if this is actually required step in debug build. Because if it's not, then it's more appropriate to leave to special occasions to remember. I mean it should not be part of default target, but one to be called explicitly. And one can also argue that target name should be easier to remember, say "docs".

**Comment by levitte** [comment]
Well, there are two ways to see it.  From a *user's* perspective, the natural thing is to `./config && make && make install` whenever there's a new release, possibly with a `make test` in the middle.  This has been repeated enough times in all kinds of projects that most people who build binaries themselves have it at their fingertips, that's why one can call it "natural" (we could say that this is the sequence of commands that everyone has been brainwashed with).

So then, the question is, why should anyone expect that `all` doesn't actually mean `all`?  Of course, this could be solved by making `build_sw` the default target instead of `all`, something I could accept, but with grudge.

Another related question is what to do if some files to install are missing.  As things currently are, any install target that can't find the files they are supposed to install will fail, and if `make install` was called, that means the full installation fails.

**Comment by dot-asm** [comment]
> From a user's perspective, the natural thing is to `./config && make && make install` whenever there's a new release

In other words what's important from user's perspective is that this **chain of commands** produces certain output. To user it doesn't really matter which operations are performed at each stage, only order. So there is no real conflict between user's and developer's perspectives. As for what to do if some files to install are missing. Well, it's really about how comprehensible error message is and when it pops up. Just make build_docs *first* install's dependency...

**Comment by levitte** [comment]
> Just make build_docs first install's dependency...

Doesn't that make this exercise quite a bit moot?  Also, for consistencies sake, I would say that `install` should also depend on `build_libs`, `build_apps` and `build_engines`.  We currently don't do that, for unknown reasons (apart for historical, we've never done that)...

**Comment by dot-asm** [comment]
> Doesn't that make this exercise quite a bit moot?

Well, as you said yourself, what if file to be installed is missing. With this in mind it would be only appropriate if install target checks if everything is in order prior proceeding with actual installation, and fail if everything is not in order. As opposite to failing installation halfway [as it supposedly is now]. So it's not meaningless to introduce target that generate docs prior installation. It's just that it doesn't have to part of default target. As for install's dependency on other build_\*, yes, it seems to be meaningful too... As for build_docs being first in the dependency, well, it simply has to be before install_\*, basically all build_\* prior install_\*...

**Comment by kaduk** [comment]
> what if file to be installed is missing

I was just running into this yesterday (attempting to ```make install_dev``` having only built a subset of (libcrypto.a, libssl.a, libcrypto.pc, libssl.pc), instead of just using ```make build_libs```).

It seems that the main obstacle to, e.g., introducing a dependency on ```build_libs``` to ```install_dev``` would be in the ```$(MAKE) depend``` step prefixing our main user-facing rules.  This recursive make invocation also causes parallel-make-unsafety, which is quite unfortunate.  (To be clear, I actually tripped over the race when doing ```make all libcrypto.a``` on a loaded system, which is not exactly a supported interface, but I believe that pessimal scheduling could cause, e.g.,  ```make build_libs build_programs``` to trigger the race as well.  The resulting build failure presented as a missing ```crypto/.../blah.d.tmp``` file, since the rule being run by the other make engine had already moved it out of the way.)  It's not clear to me that there is actual (as opposed to aesthetic) harm in running the depend step again on install, though.

**Comment by kaduk** [comment]
> From a user's perspective, the natural thing is to ./config && make && make install

I'll note without comment that we can easily make ```make``` equivalent to ```make build_sw``` as opposed to ```make all``` -- the first target in the Makefile is what's used for the default behavior.

**Comment by levitte** [comment]
I just changed things so `all` isn't the default target any more...  also made the diverse `install*` targets similar, for orthogonality.

**Comment by richsalz** [comment]
I totally agree with @dot-asm .  "make all" should build everything and "make install" should install everything.  If "all" isn't the first target, then the first target should do "make all test" or its equivalent.  I feel kinda strongly about this.

**Comment by levitte** [comment]
The default target has *never* included test, that has always been a separate target.

As this PR is right now, `make all` *does* build everything (including docs).

Frankly, the only really sane thing to do is have all the different `install` variants depend on the corresponding `build` variant.  I can do that too, but that's a subject for another PR.

**Comment by richsalz** [comment]
I strongly believe that for compatibility, "make" in 1.1.0 should be *the same* as "make" in 1.1.1

I am, frankly, a little lost with what's going on here right now.  Building the docs at, well, build time, makes sense.

**Comment by levitte** [comment]
1. you might want to take note of the label.
2. "make" in this PR currently does the same thing as in 1.1.0 or upcoming 1.1.1

(re 1: I never meant this as quick hack.  I'm *perfectly* aware that this may be a shift in habits and therefore needs time to mature...  that is the exact reason why I'm making an early WIP PR of this)

**Comment by levitte** [comment]
This has now been rebased on top of #7583, because that one provides a more direct dependency between install targets and their corresponding build targets.

I think all comments here have been met, i.e. the docs aren't built by default, you have to say `make all` explicitly to get that...  *but* the since documentation installation targets now depend on the documentation building targets, we retain the same default behaviour as before, while giving the possibility to have documentation easily built in build time.

Did I miss anything?

**Comment by levitte** [comment]
Now that #7583 is out of the way (i.e. has been merged), I did a new rebase on top of master

**Comment by levitte** [comment]
Right now, `doc/build.info` is entirely a perl snippet.  This is fine in my view, but everyone might not agree...  and it does stick out from the rest of the `build.info` files.

An options is to have it be a `build.info` file like any other and force us to add new lines every time there's a new manual written (it can be generated initially).
Another option is to have it automatically generated with a `make update`.

What say you?  Is it fine as it is or would you prefer one of the other options, and which one in that case?

**Comment by mattcaswell** [comment]
I'm ok with the build.info file as is - no need to generate it as part of `make update`

Needs a rebase.

File: Configurations/common.tmpl [hunk 1]
```diff
@@ -211,6 +219,34 @@
...
+     foreach my $doc (@{$unified_info{mandocs}->{$section}}) {
+         next if $cache{$doc};
+         $OUT .= generatesrc(src => $doc,
+                             generator => $unified_info{generate}->{$doc});
+         foreach ((@{$unified_info{depends}->{$doc}})) {
+             dogenerate($_, undef, undef, %opts);
+         }
+         $cache{$doc} = 1;
+     }
+ }
```
Comment (mattcaswell) [inline Configurations/common.tmpl]: These two functions seem virtually identical. Can we not merge them somehow?

File: doc/build.info [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+     foreach (keys %podinfiles) {
+         (my $p = $_) =~ s|\.in$||i;
+         $podfiles{$p} = 1;
+     }
+
+     foreach my $p (sort keys %podfiles) {
+         my $podfile = abs2rel($p, $sourcedir);
+	 my $podname = basename($podfile, '.pod');
+	 my $podinfile = $podinfiles{"$p.in"} ? "$podfile.in" : undef;
```
Comment (mattcaswell) [inline doc/build.info]: Indentation is strange here

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/build.info (hunk 2 above)
Comment (levitte) [reply doc/build.info]: Tabs...  and github now doing the M$ thing (4 column tabs)

File: Configurations/common.tmpl (hunk 1 above)
Comment (levitte) [reply Configurations/common.tmpl]: Sure

File: doc/build.info (hunk 2 above)
Comment (mattcaswell) [reply doc/build.info]: This still seems to be a problem

**Comment by mattcaswell** [comment]
Travis errors look relevant.

File: doc/build.info (hunk 2 above)
Comment (levitte) [reply doc/build.info]: *Now* it's done.

**Comment by levitte** [comment]
> Travis errors look relevant.

Ah, yeah, I didn't resolve the rebase conflict quite right, it seems.  Fixed, I hope

**Comment by mattcaswell** [comment]
> Fixed, I hope

Nope. Still failing.

**Comment by levitte** [comment]
Well...  at least it came a bit further this time

**Comment by levitte** [comment]
So let's see if I've pleased the CIs now

**Comment by t8m** [comment]
Unfortunately not fully yet.

**Comment by levitte** [comment]
Yeah I noticed...  I'll get back to this

**Comment by levitte** [comment]
Got it.  Chained generators weren't supported.  Now they are.

**Comment by levitte** [comment]
CIs seem happy now.  Ping?

**Review by mattcaswell — APPROVED** [review]

**Comment by levitte** [comment]
Merged.

829f86bb7b52e58c2db333b8e2eb7ad990724746 Add the possibility to generate documentation at build time
df8f116ecd6e25577eb1e47a213df17d64e87b4a Add doc/build.info to build the documentation
b0940b33a60cbfab03153287181372d0066410a6 Adapt *.tmpl to generate docs at build time
cff64af5532d3a1b0e2e9adf88454887efba01b2 Configure: make it possible to have generated generators

**Comment by levitte** [comment]
Closing this in favor of #10558, which basically does the same thing, but in source.

**Pull request closed without merging**